modules/extras.py

```diff
@@ -2,27 +2,77 @@
 import re
 import shutil
 import json
-
+import html
 
 import torch
 import tqdm
 
-from modules import shared, images, sd_models, sd_vae, sd_models_config, errors
+from modules import shared, images, sd_models, sd_vae, sd_models_config, errors, png_parser
 from modules.ui_common import plaintext_to_html
 import gradio as gr
 import safetensors.torch
 
 
+def pnginfo_format_string(plain_text):
+    content = "<br>\n".join(html.escape(x) for x in str(plain_text).split('\n'))
+    return content
+
+
+def pnginfo_format_setting(name, value):
+    cls_name = 'geninfo-setting-string' if value.startswith('"') else 'geninfo-setting-value'
+    return f"<span class='geninfo-setting-name'>{html.escape(name)}:</span> <span class='{cls_name}' onclick='uiCopyElementText(this)'>{html.escape(value)}</span>"
+
+
+def pnginfo_format_quicklink(name):
+    return f"<span class='geninfo-quick-link' onclick='uiCopyPngInfo(this, \"{name}\")'>[{html.escape(name)}]</span>"
+
+
 def run_pnginfo(image):
     if image is None:
         return '', '', ''
 
     geninfo, items = images.read_info_from_image(image)

```

Comment:
another issue: items is discarded if `parser.valid`

Reply:
@w-e-w Thank you for taking the time to go over this.  I will take the time to look over the changes and test.  It will take a few days.  Thank you again!